### PR TITLE
chore(vrl): test the VM functions as well in the `test_function` macro.

### PR DESCRIPTION
**Pull request by StephenWakely**
Ref #11195 

This expands the `test_function` macro so the VM variants of the stdlib functions are also tested.

Currently the test is skipped if the function returns `not implemented`, so all tests should pass despite not all the functions being implemented yet.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/vrl/compiler/src/value/convert.rs [hunk 1]
```diff
@@ -150,3 +151,36 @@ impl VrlValueConvert for Value {
         }
     }
 }
+
+/// Conversts from an Expr into a Value. This is only possible if the expression represents
+/// static values, Literals and Containers containing Literals.
+impl TryFrom<Expr> for Value {
```
Comment (StephenWakely) [inline lib/vrl/compiler/src/value/convert.rs:157]: @fuchsnj Is there a better place for me to put this impl?

File: lib/vrl/compiler/src/test_util.rs [hunk 2]
```diff
@@ -113,6 +113,35 @@ macro_rules! test_function {
...
+		    }
+		    err@Err(_) => {
+                        // Allow tests against compiler errors.
+                        assert_eq!(err
+                                   // We have to map to a value just to make sure the types match even though
+                                   // it will never be used.
+                                   .map(|_| Value::Null)
+                                   .map_err(|e| format!("{:#}", e.message())), want);
+                    }
+		}
```
Comment (pablosichert) [inline lib/vrl/compiler/src/test_util.rs:144]: The formatting in here got a bit wrangled 😄

File: lib/vrl/compiler/src/test_util.rs [hunk 3]
```diff
@@ -113,6 +113,35 @@ macro_rules! test_function {
                                    .map_err(|e| format!("{:#}", e.message())), want);
                     }
                 }
+
+		// Test the VM response
```
Comment (pablosichert) [inline lib/vrl/compiler/src/test_util.rs:117]:
```suggestion
		// Test the VM response.
```

File: lib/vrl/compiler/src/value/convert.rs [hunk 4]
```diff
@@ -150,3 +151,36 @@ impl VrlValueConvert for Value {
         }
     }
 }
+
+/// Converts from an Expr into a Value. This is only possible if the expression represents
+/// static values - Literals and Containers containing Literals.
```
Comment (pablosichert) [inline lib/vrl/compiler/src/value/convert.rs:156]:
```suggestion
/// Converts from an `Expr` into a `Value`. This is only possible if the expression represents
/// static values - `Literals` and `Containers` containing `Literals`.
```

File: lib/vrl/compiler/src/value/convert.rs [hunk 5]
```diff
@@ -150,3 +151,36 @@ impl VrlValueConvert for Value {
...
+            Expr::Container(Container {
+                variant: Variant::Array(array),
+            }) => Ok(Value::Array(
+                array
+                    .iter()
+                    .map(|value| value.clone().try_into())
+                    .collect::<Result<_, _>>()?,
+            )),
+            expr => Err(crate::function::Error::UnexpectedExpression {
+                keyword: "thing",
```
Comment (pablosichert) [inline lib/vrl/compiler/src/value/convert.rs:180]: What does "thing" mean in this context?

File: lib/vrl/compiler/src/test_util.rs (hunk 2 above)
Comment (StephenWakely) [reply lib/vrl/compiler/src/test_util.rs:144]: Ug.. macros...

File: lib/vrl/compiler/src/value/convert.rs (hunk 5 above)
Comment (StephenWakely) [reply lib/vrl/compiler/src/value/convert.rs:180]: Ooh good catch..

File: lib/vrl/compiler/src/test_util.rs [hunk 6]
```diff
@@ -113,6 +113,35 @@ macro_rules! test_function {
...
+                        let mut args = $crate::vm::compile_arguments(&$func, &mut args, &anys);
+                        let mut runtime_state = $crate::state::Runtime::default();
+                        let mut target: $crate::Value = ::std::collections::BTreeMap::default().into();
+                        let tz = $tz;
+                        let mut ctx = $crate::Context::new(&mut target, &mut runtime_state, &tz);
+                        let got_value_vm = $func.call_by_vm(&mut ctx, &mut args)
+                            .map_err(|e| format!("{:#}", anyhow::anyhow!(e)));
+
+                        // Don't assert results against unimplemented functions.
+                        // This needs to be removed when we implement all the functions.
```
Comment (pablosichert) [inline lib/vrl/compiler/src/test_util.rs:131]: It relatively low-priority, but an issue for this would be nice.

**Review by pablosichert — APPROVED** [review]
Nice!

**Pull request merged**